File: tutorial-scala-core/src/scala/src/main/java/com/acme/oop.java
package com.acme;

import java.util.Arrays;

interface JReportGenerator {

    byte[] generate(final String user);

}

class JReportGeneratorImpl implements JReportGenerator {
    @Override
    public byte[] generate(String user) {
        return new byte[0];
    }
}

class JReportService {
    private final JReportGenerator reportGenerator;

    public JReportService(JReportGenerator reportGenerator) {
        this.reportGenerator = reportGenerator;
    }

    void reportAll() {
        for (String user : Arrays.asList("a", "b", "c")) {
            reportGenerator.generate(user);
        }
    }
}

class JMain {
    public static void main(String[] args) {

    }
}

File: tutorial-scala-core/src/scala/src/main/scala/com/acme/oop.scala
package com.acme

trait ReportGenerator {
  def generate(user: String): Array[Byte]
}

class ReportGeneratorImpl extends ReportGenerator with Cloneable {
  override def generate(user: String): Array[Byte] = Array.emptyByteArray
  override def clone(): AnyRef = ???
}

class ReportService(reportGenerator: ReportGenerator) {

  def reportAll(): Unit = {
    for (user ← Seq("a", "b", "c")) {
      reportGenerator.generate(user)
    }
  }
}

object Main extends App {

  val rs = new ReportService(new ReportGeneratorImpl)
  rs.reportAll()

}
